File: Backend/notifications/noti_model.py
from datetime import date, time, datetime
from uuid import UUID
from typing import Union, Literal, Annotated, List

from pydantic import BaseModel, Field, field_serializer, model_validator


# ------------------------------------------------------------------
# Shared base
# ------------------------------------------------------------------
class NotificationSchedule(BaseModel):
    schedule_type: Literal["date", "weekly"]
    times: List[time]
    timezone: str = "UTC"
    is_active: bool = True

    scheduled_date: date | None = None
    weekdays: List[Annotated[int, Field(ge=0, le=6)]] | None = None

    @field_serializer("scheduled_date")
    def serialize_date(self, value: date | None):
        return value.isoformat() if value else None

    @field_serializer("times")
    def serialize_times(self, times: List[time]):
        return [t.isoformat() for t in times]

    #? Checks notification type and removes the unnecessary field and also checks if the required field is present
    @model_validator(mode="after")
    def normalize_schedule_fields(self):
        if self.schedule_type == "weekly":
            self.scheduled_date = None
            if not self.weekdays:
                raise ValueError("weekdays is required for weekly notifications")
        else:  # "date"
            self.weekdays = None
            if self.scheduled_date is None:
                raise ValueError("scheduled_date is required for date notifications")
        return self


class NotificationCreate(NotificationSchedule):
    task_id: int


class NotificationRead(NotificationSchedule):
    id: UUID
    task_id: int
    created_at: datetime

File: Frontend/lib/Data/Models/notiModel.dart
class NotificationModel {
  int? taskId;
  final String scheduleType; // "date" or "weekly"
  final List<String> times;
  final String timezone;
  final bool isActive;

  // optional fields
  final String? scheduledDate; // only for "date"
  final List<int>? weekdays; // only for "weekly"

  NotificationModel({
    this.taskId,
    required this.scheduleType,
    required this.times,
    required this.timezone,
    required this.isActive,
    this.scheduledDate,
    this.weekdays,
  });

  factory NotificationModel.fromJson(Map<String, dynamic> json) {
    return NotificationModel(
      taskId: json['task_id'],
      scheduleType: json['schedule_type'],
      times: List<String>.from(json['times']),
      timezone: json['timezone'],
      isActive: json['is_active'],
      scheduledDate: json['scheduled_date'],
      weekdays: json['weekdays'] != null
          ? List<int>.from(json['weekdays'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      'task_id': taskId,
      'schedule_type': scheduleType,
      'times': times,
      'timezone': timezone,
      'is_active': isActive,
    };

    if (scheduleType == 'date') {
      data['scheduled_date'] = scheduledDate;
    }

    if (scheduleType == 'weekly') {
      data['weekdays'] = weekdays;
    }

    return data;
  }

  NotificationModel copyWith({
    int? taskId,
    String? scheduleType,
    List<String>? times,
    String? timezone,
    bool? isActive,
    String? scheduledDate,
    List<int>? weekdays,
  }) {
    return NotificationModel(
      taskId: taskId ?? this.taskId,
      scheduleType: scheduleType ?? this.scheduleType,
      times: times ?? this.times,
      timezone: timezone ?? this.timezone,
      isActive: isActive ?? this.isActive,
      scheduledDate: scheduledDate ?? this.scheduledDate,
      weekdays: weekdays ?? this.weekdays,
    );
  }
}
